File: src/ClaudeUsageWatcher/UsageService.cs
using System.Net.Http;
using ClaudeUsageWatcher.Core;

namespace ClaudeUsageWatcher;

public enum FetchOutcome
{
    Success,
    NoCredentials,
    AuthFailed,
    RateLimited,
    Error
}

public sealed record FetchResult(FetchOutcome Outcome, UsageSnapshot? Snapshot, string? Message)
{
    public static FetchResult Success(UsageSnapshot snapshot) =>
        new(FetchOutcome.Success, snapshot, null);

    public static FetchResult NoCredentials() =>
        new(FetchOutcome.NoCredentials, null, null);

    public static FetchResult AuthFailed() =>
        new(FetchOutcome.AuthFailed, null, null);

    public static FetchResult RateLimited() =>
        new(FetchOutcome.RateLimited, null, null);

    public static FetchResult Error(string message) =>
        new(FetchOutcome.Error, null, message);
}

public sealed class UsageService : IDisposable
{
    private readonly HttpClient _httpClient;
    private readonly UsageClient _usageClient;
    private readonly CredentialsReader _credentialsReader;

    public UsageSnapshot? LastGood { get; private set; }
    public bool IsStale { get; private set; }
    public bool HasCredentials { get; private set; } = true;

    public UsageService()
    {
        _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };
        _usageClient = new UsageClient(_httpClient);
        _credentialsReader = new CredentialsReader();
    }

    public async Task<FetchResult> PollAsync(CancellationToken ct = default)
    {
        ClaudeCredentials? creds = _credentialsReader.TryRead();
        if (creds is null || string.IsNullOrEmpty(creds.AccessToken))
        {
            HasCredentials = false;
            IsStale = LastGood is not null;
            return FetchResult.NoCredentials();
        }

        HasCredentials = true;

        try
        {
            UsageSnapshot snapshot = await _usageClient
                .FetchAsync(creds.AccessToken, ct)
                .ConfigureAwait(false);

            LastGood = snapshot;
            IsStale = false;
            return FetchResult.Success(snapshot);
        }
        catch (UsageAuthException)
        {
            IsStale = LastGood is not null;
            return FetchResult.AuthFailed();
        }
        catch (UsageRateLimitException)
        {
            IsStale = LastGood is not null;
            return FetchResult.RateLimited();
        }
        catch (OperationCanceledException)
        {
            IsStale = LastGood is not null;
            return FetchResult.Error("cancelled");
        }
        catch (Exception ex)
        {
            IsStale = LastGood is not null;
            return FetchResult.Error(ex.Message);
        }
    }

    public void Dispose()
    {
        _httpClient.Dispose();
    }
}

File: macos/Sources/UsageCore/UsageService.swift
import Foundation

public enum FetchOutcome: Sendable {
    case success
    case noCredentials
    case authFailed
    case rateLimited
    case error(String)
}

public struct FetchResult: Sendable {
    public let outcome: FetchOutcome
    public let snapshot: UsageSnapshot?
    public let notification: ThresholdEvent?

    public init(outcome: FetchOutcome, snapshot: UsageSnapshot?, notification: ThresholdEvent?) {
        self.outcome = outcome
        self.snapshot = snapshot
        self.notification = notification
    }
}

/// Poll orchestration mirroring the C# UsageService: reads credentials, fetches a
/// snapshot, maps errors to outcomes, and maintains last-good / staleness state.
public final class UsageService {
    private let usageClient: UsageClient
    private let credentialsReader: CredentialsReader
    private let thresholdTracker: ThresholdTracker
    private let burnRate = BurnRateEstimator()

    public private(set) var lastGood: UsageSnapshot?
    public private(set) var isStale = false
    public private(set) var hasCredentials = true

    public init(
        notifyWarnAt: Double = 80,
        notifyCriticalAt: Double = 95,
        usageClient: UsageClient? = nil,
        credentialsReader: CredentialsReader? = nil
    ) {
        self.usageClient = usageClient ?? UsageClient(transport: URLSessionTransport())
        self.credentialsReader = credentialsReader ?? CredentialsReader()
        self.thresholdTracker = ThresholdTracker(warnAt: notifyWarnAt, criticalAt: notifyCriticalAt)
    }

    public func projectTimeToCap(now: Date) -> TimeInterval? {
        burnRate.projectTimeToCap(now: now)
    }

    public func poll() async -> FetchResult {
        guard let creds = credentialsReader.tryRead(), !creds.accessToken.isEmpty else {
            hasCredentials = false
            isStale = lastGood != nil
            return FetchResult(outcome: .noCredentials, snapshot: nil, notification: nil)
        }

        hasCredentials = true

        do {
            let snapshot = try await usageClient.fetch(accessToken: creds.accessToken)
            lastGood = snapshot
            isStale = false

            var notification: ThresholdEvent?
            if let five = snapshot.fiveHour {
                notification = thresholdTracker.observe(five.utilization)
                burnRate.add(at: snapshot.fetchedAt, utilization: five.utilization)
            }

            return FetchResult(outcome: .success, snapshot: snapshot, notification: notification)
        } catch UsageError.auth {
            isStale = lastGood != nil
            return FetchResult(outcome: .authFailed, snapshot: nil, notification: nil)
        } catch UsageError.rateLimited {
            isStale = lastGood != nil
            return FetchResult(outcome: .rateLimited, snapshot: nil, notification: nil)
        } catch {
            isStale = lastGood != nil
            return FetchResult(outcome: .error(error.localizedDescription), snapshot: nil, notification: nil)
        }
    }
}
